pkg/volume/csi/csi_block.go

```diff
@@ -331,31 +373,96 @@ func (m *csiBlockMapper) TearDownDevice(globalMapPath, devicePath string) error
 	}
 
-	// Call NodeUnstageVolume
-	stagingPath := m.getStagingPath()
-	if _, err := os.Stat(stagingPath); err != nil {
+	dataDir := getVolumeDeviceDataDir(m.specName, m.plugin.host)
+
+	// remove ~/${pv}/data/vol_data.json first, then remove the other dir.
+	volDataFile := path.Join(dataDir, volDataFileName)
+	err = os.Remove(volDataFile)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	// remove ~/${pv}/data
+	err = os.Remove(dataDir)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	// remove ~/${pv}
+	pvPath := filepath.Dir(dataDir)
+	err = os.Remove(pvPath)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	return nil
+}
+
+// UnmapPodDevice unmaps the block device path.
+func (m *csiBlockMapper) UnmapPodDevice(globalMapPath, volumeMapPath, volumeMapName string, podUID types.UID) error {
+	if !m.plugin.blockEnabled {
+		return errors.New("CSIBlockVolume feature not enabled")
+	}
+
+	klog.V(4).Infof(log("unmapper.UnmapPodDevice(globalMapPath=%s, volumeMapPath,=%s, volumeMapName=%s, podUID=%s)",
+		globalMapPath, volumeMapPath, volumeMapName, string(podUID)))
+
+	err := ioutil.UnmapBlockVolume(globalMapPath, volumeMapPath, volumeMapName, podUID)
+	if err != nil {
+		return err
+	}
+
+	publishPath := m.getPublishPath()
+	blkUtil := volumepathhandler.NewBlockVolumePathHandler()
+	// Search under globalMapPath dir if all symbolic links from pods have been removed already.
+	// If symbolic links are there, pods may still refer the volume.
+	refs, err := blkUtil.GetDeviceSymlinkRefs(publishPath, globalMapPath)
+	if err != nil {
+		return err
+	}
+	if len(refs) > 0 {
+		return fmt.Errorf("blockMapper.UnmapPodDevice The device %q is still referenced from other Pods %v", globalMapPath, refs)
+	}
+
+	// The block volume is not referenced from Pods. Release file descriptor lock.
+	// This should be done before calling UnmapPodDevice, because some plugins that do local detach
+	// in UnmapDownDevice will fail in detaching device due to the refcnt on the loopback device.
+	err = blkUtil.DetachFileDevice(publishPath)
+	if err != nil {
+		return fmt.Errorf("blockMapper.UnmapPodDevice DetachFileDevice failed: %v", err)
+	}
+
+	csiClient, err := m.csiClientGetter.Get()
+	if err != nil {
+		return errors.New(log("blockMapper.UnmapPodDevice failed to get CSI client: %v", err))
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), csiTimeout)
+	defer cancel()
+
+	// Call NodeUnpublishVolume
+	if _, err := os.Stat(publishPath); err != nil {
 		if os.IsNotExist(err) {
-			klog.V(4).Infof(log("blockMapper.TearDownDevice stagingPath(%s) has already been deleted, skip calling NodeUnstageVolume", stagingPath))
+			klog.V(4).Infof(log("blockMapper.UnmapPodDevice publishPath(%s) has already been deleted, skip calling NodeUnpublishVolume", publishPath))
 		} else {
 			return err
 		}
 	} else {
-		err := m.unstageVolumeForBlock(ctx, csiClient, stagingPath)
+		err := m.unpublishVolumeForBlock(ctx, csiClient, publishPath)
```

Comment:
since we have changed publishPath and stagingPath, there might be a backwards-compatibility issue.

Previously, we created a symlink in the pod's volume directory to `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}`.
Now, we create a symlink in the pod's volume directory to `plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}`.

After an upgrade, we ask the plugin to unpublish `plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}` even if it, before the upgrade, published `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}`. So `plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}` will be stuck in published state.

We may need to follow the symlink in the pod's volume directory to see if it links to the new path or the old path and call unpublish appropriately.

Reply:
proper upgrade procedure you are supposed to drain the node first. So I'm not sure this is worth fixing.

Reply:
Good investigation!
Then, should we add a caution like below in release notes?

```
Before this release, staging path and publish path for CSI block volume were as follows:
- Staging path: plugins/kubernetes.io/csi/volumeDevices/staging/{pvname}
- Publish path: plugins/kubernetes.io/csi/volumeDevices/publish/{pvname}
In this release these are changed as follows:
- Staging path: plugins/kubernetes.io/csi/volumeDevices/staging/{volumeID}
- Publish path: plugins/kubernetes.io/csi/volumeDevices/publish/{volumeID}/{podUID}

As a result, upgrading Kubernetes cluster while any pods with CSI block volumes running on a node will cause these path inconsistent, which will lead failure in unstaging/unpublishing volume. 
Therefore, all nodes need to be drained before upgrading Kubernetes cluster.
```

Any improvements in wording are welcome.

Reply:
Thanks, IMO it doesn't need to be too detailed about the specific paths, the last two sentences reminding people that nodes need to be drained else unpublish/unstage will fail should be enough.